File: src/main/java/com/lczq/dbquery/util/SignUtil.java
package com.lczq.dbquery.util;

import com.lczq.dbquery.entities.SignEntity;
import org.springframework.util.DigestUtils;

import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.Random;

public class SignUtil
{
    public static SignEntity generateSign()
    {
        SignEntity signEntity = new SignEntity();
        byte[] array = new byte[32];
        new Random().nextBytes(array);
        String appId = DigestUtils.md5DigestAsHex(array).substring(0,16);
        new Random().nextBytes(array);
        String appKey = DigestUtils.md5DigestAsHex(array);
        signEntity.setAppId(appId);
        signEntity.setAppKey(appKey);
        return signEntity;
    }

    public static boolean validSign(String sign, Map<String, String> queryParams, SignEntity signEntity)
    {
        // sort real params
        String sb = ParamUtil.sortedParams(queryParams) + signEntity.toString();
        String validSign = DigestUtils.md5DigestAsHex(sb.getBytes(StandardCharsets.UTF_8));
        return validSign.equals(sign);

    }
}

File: backend/src/main/kotlin/com/github/wgzhao/dbquery/util/SignUtil.kt
package com.github.wgzhao.dbquery.util

import com.github.wgzhao.dbquery.entities.Sign
import org.springframework.util.DigestUtils
import java.nio.charset.StandardCharsets
import java.util.*

object SignUtil {
    fun generateSign(): Sign {
        val array = ByteArray(32)
        Random().nextBytes(array)
        val appId = DigestUtils.md5DigestAsHex(array).substring(0, 16)
        Random().nextBytes(array)
        val appKey = DigestUtils.md5DigestAsHex(array)
        return Sign(appId, appKey)
    }

    fun validSign(sign: String?, queryParams: MutableMap<String, String>?, signs: Sign): Boolean {
        // sort real params
        if (queryParams.isNullOrEmpty()) {
            return sign.isNullOrEmpty()
        }
        val sb = ParamUtil.sortedParams(queryParams) + signs.toString()
        val validSign = DigestUtils.md5DigestAsHex(sb.toByteArray(StandardCharsets.UTF_8))
        return validSign == sign
    }
}
